File: node-sidecar-test/test/servidor.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';
import * as C from '../src/constantes.js';
import type { Motor, Salida } from '../src/ejecucion.js';
import { Servidor } from '../src/servidor.js';
import { pedir, pedirCrudo, rutaSocket } from './ayudas.js';

/** Secciones 3 y 9: el contrato HTTP hacia el worker y la saturacion. */

const ID = '3f2b1c4d-0000-4000-8000-000000000001';

type Respuesta = Awaited<ReturnType<typeof pedir>>;

/** Motor de mentira: cuenta llamadas y se puede dejar colgado para llenar la cola. */
class MotorDePrueba implements Motor {
  llamadas = 0;
  ultimoTar: Buffer | null = null;
  daemonOk = true;
  private liberar: (() => void) | null = null;
  private readonly bloqueado: Promise<void>;

  constructor(colgado = false) {
    this.bloqueado = colgado ? new Promise((r) => { this.liberar = r; }) : Promise.resolve();
  }

  async ejecutar(ejecucionId: string, tar: Buffer): Promise<Salida> {
    this.llamadas++;
    this.ultimoTar = tar;
    await this.bloqueado;
    return {
      ejecucionId, resultado: 'COMPLETADA', exitCode: 0, oomKilled: false, duracionMs: 1,
      stdout: '', stderr: '', reporte: null, reporteAusente: true, salidaTruncada: false,
    };
  }

  daemonVivo(): Promise<boolean> { return Promise.resolve(this.daemonOk); }
  limpiarEnVuelo(): Promise<void> { return Promise.resolve(); }
  soltar(): void { this.liberar?.(); }
}

async function conServidor<T>(motor: Motor, cuerpo: (ruta: string) => Promise<T>): Promise<T> {
  const ruta = rutaSocket('ejecutor');
  const servidor = new Servidor(ruta, motor);
  await servidor.escuchar();
  try {
    return await cuerpo(ruta);
  } finally {
    await servidor.cerrar();
  }
}

const cabeceras = (largo: number) => ({ 'X-Ejecucion-Id': ID, 'Content-Length': String(largo) });

test('el camino feliz devuelve 200 con el json de la seccion 3.1', async () => {
  const motor = new MotorDePrueba();
  await conServidor(motor, async (ruta) => {
    const r = await pedir(ruta, {
      metodo: 'POST', ruta: '/ejecutar', cabeceras: cabeceras(3), cuerpo: Buffer.from('tar'),
    });
    assert.equal(r.codigo, 200);
    const salida = JSON.parse(r.cuerpo);
    assert.deepEqual(Object.keys(salida), ['ejecucionId', 'resultado', 'exitCode', 'oomKilled',
      'duracionMs', 'stdout', 'stderr', 'reporte', 'reporteAusente', 'salidaTruncada']);
    assert.deepEqual(motor.ultimoTar, Buffer.from('tar'));
  });
});

test('400 si falta X-Ejecucion-Id o no es un uuid', async () => {
  const motor = new MotorDePrueba();
  await conServidor(motor, async (ruta) => {
    const sinId = await pedir(ruta, { metodo: 'POST', ruta: '/ejecutar', cabeceras: { 'Content-Length': '0' } });
    assert.equal(sinId.codigo, 400);
    const malId = await pedir(ruta, {
      metodo: 'POST', ruta: '/ejecutar', cabeceras: { 'X-Ejecucion-Id': 'no-uuid', 'Content-Length': '0' },
    });
    assert.equal(malId.codigo, 400);
    assert.equal(motor.llamadas, 0);
  });
});

test('el id no puede traer barras ni query: el nombre del contenedor se concatena en la url', async () => {
  const motor = new MotorDePrueba();
  await conServidor(motor, async (ruta) => {
    const r = await pedir(ruta, {
      metodo: 'POST',
      ruta: '/ejecutar',
      cabeceras: { 'X-Ejecucion-Id': '3f2b1c4d-0000-4000-8000-000000000001/../x', 'Content-Length': '0' },
    });
    assert.equal(r.codigo, 400);
    assert.equal(motor.llamadas, 0);
  });
});

test('411 si no viene Content-Length: no se acepta chunked en la entrada', async () => {
  const motor = new MotorDePrueba();
  await conServidor(motor, async (ruta) => {
    // Necesitamos conocer el tamano antes de aceptar bytes, asi que un cuerpo chunked se rechaza.
    const r = await pedirCrudo(ruta, `POST /ejecutar HTTP/1.1\r\nHost: localhost\r\n`
      + `X-Ejecucion-Id: ${ID}\r\nTransfer-Encoding: chunked\r\nConnection: close\r\n\r\n`
      + `3\r\ntar\r\n0\r\n\r\n`);
    assert.match(r, /^HTTP\/1\.1 411 /);
    assert.equal(motor.llamadas, 0);
  });
});

test('a27 413 con un bundle mayor al tope, sin leer el cuerpo', async () => {
  const motor = new MotorDePrueba();
  await conServidor(motor, async (ruta) => {
    // Se declaran MAX_BUNDLE_BYTES + 1 y se mandan solo unos pocos bytes: si el servidor esperara
    // el cuerpo completo esto no responderia nunca.
    const r = await pedir(ruta, {
      metodo: 'POST', ruta: '/ejecutar',
      cabeceras: cabeceras(C.MAX_BUNDLE_BYTES + 1), cuerpo: Buffer.alloc(16),
    });
    assert.equal(r.codigo, 413);
    assert.equal(motor.llamadas, 0);
  });
});

test('el bundle de exactamente MAX_BUNDLE_BYTES se acepta', async () => {
  const motor = new MotorDePrueba();
  await conServidor(motor, async (ruta) => {
    const r = await pedir(ruta, {
      metodo: 'POST', ruta: '/ejecutar',
      cabeceras: cabeceras(C.MAX_BUNDLE_BYTES), cuerpo: Buffer.alloc(C.MAX_BUNDLE_BYTES, 0x41),
    });
    assert.equal(r.codigo, 200);
    assert.equal(motor.ultimoTar!.length, C.MAX_BUNDLE_BYTES);
  });
});

/** Lanza `cuantos` pedidos y espera a que el servidor los tenga contados: en vuelo mas en cola. */
async function saturar(ruta: string, cuantos: number): Promise<Array<Promise<Respuesta>>> {
  const pedidos: Array<Promise<Respuesta>> = [];
  for (let i = 0; i < cuantos; i++) {
    pedidos.push(pedir(ruta, {
      metodo: 'POST', ruta: '/ejecutar', cabeceras: cabeceras(1), cuerpo: Buffer.from('x'),
    }));
  }
  // Sin esta espera la carrera seria entre el cliente y el event loop, no entre los pedidos.
  for (let intento = 0; intento < 500; intento++) {
    const salud = JSON.parse((await pedir(ruta, { metodo: 'GET', ruta: '/salud' })).cuerpo);
    if (salud.enVuelo + salud.enCola >= cuantos) return pedidos;
    await new Promise((r) => setTimeout(r, 5));
  }
  throw new Error('el servidor no llego a admitir todos los pedidos');
}

test('a26 con MAX_CONCURRENTES + MAX_COLA + 1 pedidos simultaneos exactamente uno recibe 503', async () => {
  const motor = new MotorDePrueba(true);
  await conServidor(motor, async (ruta) => {
    const admitidos = await saturar(ruta, C.MAX_CONCURRENTES + C.MAX_COLA);
    // Solo MAX_CONCURRENTES llegan al motor; el resto espera turno (R9.1).
    assert.equal(motor.llamadas, C.MAX_CONCURRENTES);

    // El que sobra: por encima de MAX_CONCURRENTES + MAX_COLA, 503 inmediato (R9.2).
    const sobrante = await pedir(ruta, {
      metodo: 'POST', ruta: '/ejecutar', cabeceras: cabeceras(1), cuerpo: Buffer.from('x'),
    });
    assert.equal(sobrante.codigo, 503);
    assert.equal(sobrante.cabeceras['retry-after'], '5');
    assert.equal(JSON.parse(sobrante.cuerpo).resultado, 'RECHAZADA');

    motor.soltar();
    const respuestas = await Promise.all(admitidos);
    assert.equal(respuestas.filter((r) => r.codigo === 200).length, C.MAX_CONCURRENTES + C.MAX_COLA);
  });
});

test('r3.3 con la cola llena /salud sigue devolviendo 200', async () => {
  const motor = new MotorDePrueba(true);
  await conServidor(motor, async (ruta) => {
    const admitidos = await saturar(ruta, C.MAX_CONCURRENTES + C.MAX_COLA);
    const salud = await pedir(ruta, { metodo: 'GET', ruta: '/salud' });
    assert.equal(salud.codigo, 200, 'saturacion no es enfermedad');
    const cuerpo = JSON.parse(salud.cuerpo);
    assert.equal(cuerpo.daemon, 'ok');
    assert.equal(cuerpo.enVuelo, C.MAX_CONCURRENTES);
    assert.equal(cuerpo.enCola, C.MAX_COLA);

    motor.soltar();
    await Promise.all(admitidos);
  });
});

test('a29 /salud devuelve 503 si el daemon no responde', async () => {
  const motor = new MotorDePrueba();
  motor.daemonOk = false;
  await conServidor(motor, async (ruta) => {
    const r = await pedir(ruta, { metodo: 'GET', ruta: '/salud' });
    assert.equal(r.codigo, 503);
    assert.equal(JSON.parse(r.cuerpo).daemon, 'caido');
  });
});

test('un ERROR_DAEMON del motor se responde con 502', async () => {
  const motor: Motor = {
    ejecutar: async (ejecucionId) => ({
      ejecucionId, resultado: 'ERROR_DAEMON', exitCode: null, oomKilled: false, duracionMs: 4,
      stdout: '', stderr: '', reporte: null, reporteAusente: true, salidaTruncada: false,
    }),
    daemonVivo: async () => true,
    limpiarEnVuelo: async () => undefined,
  };
  await conServidor(motor, async (ruta) => {
    const r = await pedir(ruta, { metodo: 'POST', ruta: '/ejecutar', cabeceras: cabeceras(0) });
    assert.equal(r.codigo, 502);
  });
});

test('un TIMEOUT se responde con 200: es un resultado, no un error', async () => {
  const motor: Motor = {
    ejecutar: async (ejecucionId) => ({
      ejecucionId, resultado: 'TIMEOUT', exitCode: null, oomKilled: false, duracionMs: 60000,
      stdout: 'parcial', stderr: '', reporte: null, reporteAusente: true, salidaTruncada: true,
    }),
    daemonVivo: async () => true,
    limpiarEnVuelo: async () => undefined,
  };
  await conServidor(motor, async (ruta) => {
    const r = await pedir(ruta, { metodo: 'POST', ruta: '/ejecutar', cabeceras: cabeceras(0) });
    assert.equal(r.codigo, 200);
    assert.equal(JSON.parse(r.cuerpo).resultado, 'TIMEOUT');
  });
});

test('no existe ningun otro endpoint', async () => {
  const motor = new MotorDePrueba();
  await conServidor(motor, async (ruta) => {
    for (const prueba of ['/ejecutar/otro', '/containers/create', '/']) {
      assert.equal((await pedir(ruta, { metodo: 'GET', ruta: prueba })).codigo, 404);
    }
    // R2.2 / no-requerimiento: no hay forma de elegir imagen, limites, red ni comando.
    assert.equal((await pedir(ruta, { metodo: 'PUT', ruta: '/ejecutar' })).codigo, 404);
  });
});

test('a30 r11.7 tras cerrar, el socket deja de aceptar', async () => {
  const motor = new MotorDePrueba();
  const ruta = rutaSocket('ejecutor');
  const servidor = new Servidor(ruta, motor);
  await servidor.escuchar();
  assert.equal((await pedir(ruta, { metodo: 'GET', ruta: '/salud' })).codigo, 200);
  await servidor.cerrar();
  await assert.rejects(pedir(ruta, { metodo: 'GET', ruta: '/salud' }));
});

File: ms-sandbox/ejecutor/src/test/java/sandbox/ejecutor/ServidorTest.java
package sandbox.ejecutor;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.*;

/** Seccion 3 y seccion 9: contrato HTTP, saturacion y limites de entrada. */
class ServidorTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    /** El catalogo de esta suite: un solo perfil, la misma clave que manda ClienteHttpDePrueba por defecto. */
    private static final Catalogo CATALOGO = Catalogo.deUnSolo(Catalogo.Perfil.armar(
            "perfil-prueba", 1, "sandbox-runner:prueba",
            "#!/bin/sh\necho prueba\n".getBytes(java.nio.charset.StandardCharsets.UTF_8),
            "junit-xml", 512, 20));

    /** Motor de mentira: no toca Docker, y puede quedarse trabado a pedido para llenar la cola. */
    private static final class MotorDePrueba implements Motor {
        final CountDownLatch soltar = new CountDownLatch(1);
        final AtomicInteger enEjecucion = new AtomicInteger();
        volatile boolean bloquear = false;
        volatile boolean daemonVivo = true;
        volatile boolean limpiado = false;

        @Override
        public Ejecucion.Salida ejecutar(String id, byte[] tar, String perfilClave) {
            enEjecucion.incrementAndGet();
            try {
                if (bloquear) soltar.await();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                enEjecucion.decrementAndGet();
            }
            return new Ejecucion.Salida(id, Ejecucion.Estado.COMPLETADA, 0, false, 12,
                    "salida", "", "<testsuite/>", false, false,
                    "perfil-prueba", 1, "hash-de-prueba");
        }

        @Override public boolean daemonVivo() { return daemonVivo; }
        @Override public void limpiarEnVuelo() { limpiado = true; }
    }

    private Path directorio;
    private String rutaSocket;
    private Servidor servidor;
    private ExecutorService hilos;
    private MotorDePrueba motor;
    private ClienteHttpDePrueba cliente;

    @BeforeEach
    void levantar() throws IOException {
        directorio = Files.createTempDirectory("ej");
        rutaSocket = directorio.resolve("e.sock").toString();
        motor = new MotorDePrueba();
        hilos = Executors.newVirtualThreadPerTaskExecutor();
        servidor = new Servidor(rutaSocket, motor, CATALOGO, hilos);
        hilos.submit(servidor::atender);
        cliente = new ClienteHttpDePrueba(rutaSocket);
    }

    @AfterEach
    void bajar() throws IOException {
        motor.soltar.countDown();
        servidor.close();
        hilos.shutdownNow();
        try (var f = Files.walk(directorio)) {
            f.sorted(java.util.Comparator.reverseOrder()).forEach(p -> p.toFile().delete());
        }
    }

    private static String id() { return UUID.randomUUID().toString(); }

    @Test
    void caminoFelizDevuelveElContratoDeLaSeccion3() throws Exception {
        var r = cliente.ejecutar(id(), "tar de mentira".getBytes());
        assertEquals(200, r.codigo());
        JsonNode n = MAPPER.readTree(r.cuerpo());
        assertEquals("COMPLETADA", n.get("resultado").asText());
        assertEquals(0, n.get("exitCode").asInt());
        assertEquals("salida", n.get("stdout").asText());
        assertEquals("<testsuite/>", n.get("reporte").asText());
        assertFalse(n.get("oomKilled").asBoolean());
        assertFalse(n.get("reporteAusente").asBoolean());
        assertFalse(n.get("salidaTruncada").asBoolean());
        assertTrue(n.has("duracionMs"));
    }

    /**
     * La serializacion de la respuesta se apoya en que Jackson emite los componentes del record
     * en orden de declaracion. Este test fija ese contrato: los diez campos de la seccion 3.1,
     * en ese orden, con los nulos presentes en vez de omitidos, EXTENDIDO (Paso 2 del handoff) con
     * perfilId/perfilVersion/perfilHash del catalogo de perfiles al final.
     */
    @Test
    void laRespuestaTieneLosCamposDeLaSeccion3EnOrden() {
        var timeout = new Ejecucion.Salida("una-id", Ejecucion.Estado.TIMEOUT, null, true, 60001,
                "parcial", "", null, true, true, null, null, null);
        assertEquals("{\"ejecucionId\":\"una-id\",\"resultado\":\"TIMEOUT\",\"exitCode\":null,"
                + "\"oomKilled\":true,\"duracionMs\":60001,\"stdout\":\"parcial\",\"stderr\":\"\","
                + "\"reporte\":null,\"reporteAusente\":true,\"salidaTruncada\":true,"
                + "\"perfilId\":null,\"perfilVersion\":null,\"perfilHash\":null}", Servidor.json(timeout));
    }

    @Test
    void errorDaemonEs502() throws Exception {
        Servidor otro = null;
        try {
            Motor caido = new Motor() {
                @Override public Ejecucion.Salida ejecutar(String id, byte[] tar, String perfilClave) {
                    return Ejecucion.Salida.errorDaemon(id, 3);
                }
                @Override public boolean daemonVivo() { return false; }
                @Override public void limpiarEnVuelo() {}
            };
            String ruta = directorio.resolve("d.sock").toString();
            otro = new Servidor(ruta, caido, CATALOGO, hilos);
            hilos.submit(otro::atender);

            var r = new ClienteHttpDePrueba(ruta).ejecutar(id(), new byte[0]);
            assertEquals(502, r.codigo());
            assertEquals("ERROR_DAEMON", MAPPER.readTree(r.cuerpo()).get("resultado").asText());
        } finally {
            if (otro != null) otro.close();
        }
    }

    /** Seccion 3.3: falta el id, o no es un UUID. */
    @Test
    void sinIdValidoEs400() throws Exception {
        var sinHeader = cliente.enviar("POST /ejecutar HTTP/1.1\r\nHost: localhost\r\nContent-Length: 0\r\n\r\n", null);
        assertEquals(400, sinHeader.codigo());

        var malFormado = cliente.ejecutar("no-soy-un-uuid", new byte[0]);
        assertEquals(400, malFormado.codigo());
    }

    /** El id se concatena en la URL de create: un id con barras seria una inyeccion. */
    @Test
    void idConCaracteresDeRutaEs400() throws Exception {
        assertEquals(400, cliente.ejecutar("../../containers/otro", new byte[0]).codigo());
    }

    /** Paso 2 del handoff (catalogo de perfiles): X-Perfil ausente o con formato invalido es 400. */
    @Test
    void sinPerfilValidoEs400() throws Exception {
        var sinHeader = cliente.enviar("POST /ejecutar HTTP/1.1\r\nHost: localhost\r\n"
                + "X-Ejecucion-Id: " + id() + "\r\nContent-Length: 0\r\n\r\n", null);
        assertEquals(400, sinHeader.codigo());

        var malFormado = cliente.enviar("POST /ejecutar HTTP/1.1\r\nHost: localhost\r\n"
                + "X-Ejecucion-Id: " + id() + "\r\nX-Perfil: no-tiene-arroba\r\n"
                + "Content-Length: 0\r\n\r\n", null);
        assertEquals(400, malFormado.codigo());
    }

    /** Paso 2 del handoff: header bien formado pero ausente del catalogo es 422. */
    @Test
    void perfilDesconocidoEs422() throws Exception {
        var r = cliente.ejecutar(id(), new byte[0], "no-existe@99");
        assertEquals(422, r.codigo());
        assertEquals(0, motor.enEjecucion.get());
    }

    /** Seccion 3.1: no se acepta entrada sin Content-Length. */
    @Test
    void sinContentLengthEs411() throws Exception {
        var r = cliente.enviar("POST /ejecutar HTTP/1.1\r\nHost: localhost\r\n"
                + "X-Ejecucion-Id: " + id() + "\r\n"
                + "X-Perfil: " + ClienteHttpDePrueba.PERFIL_POR_DEFECTO + "\r\n"
                + "Transfer-Encoding: chunked\r\n\r\n", null);
        assertEquals(411, r.codigo());
    }

    /**
     * A27: bundle de MAX_BUNDLE_BYTES + 1 responde 413 sin haber leido el cuerpo.
     * El test declara el largo y NO manda un solo byte: si el servidor intentara leerlo, se colgaria.
     */
    @Test
    @Timeout(10)
    void a27_bundleDemasiadoGrandeEs413SinLeerElCuerpo() throws Exception {
        var r = cliente.enviar("POST /ejecutar HTTP/1.1\r\nHost: localhost\r\n"
                + "X-Ejecucion-Id: " + id() + "\r\n"
                + "X-Perfil: " + ClienteHttpDePrueba.PERFIL_POR_DEFECTO + "\r\n"
                + "Content-Length: " + (Constantes.MAX_BUNDLE_BYTES + 1) + "\r\n\r\n", null);
        assertEquals(413, r.codigo());
        assertEquals(0, motor.enEjecucion.get());
    }

    @Test
    void unBundleJustoEnElLimiteSeAcepta() throws Exception {
        var r = cliente.ejecutar(id(), new byte[Constantes.MAX_BUNDLE_BYTES]);
        assertEquals(200, r.codigo());
    }

    /**
     * A26: MAX_CONCURRENTES + MAX_COLA + 1 pedidos simultaneos, exactamente uno recibe 503.
     * Se espera a que /salud muestre la cola llena antes de mandar el ultimo, para que el test
     * afirme el limite y no una carrera.
     */
    @Test
    @Timeout(60)
    void a26_conLaColaLlenaExactamenteUnoRecibe503() throws Exception {
        motor.bloquear = true;
        int admitidos = Constantes.MAX_CONCURRENTES + Constantes.MAX_COLA;

        List<Future<Integer>> enVuelo = new ArrayList<>();
        for (int i = 0; i < admitidos; i++) {
            enVuelo.add(hilos.submit(() -> cliente.ejecutar(id(), new byte[8]).codigo()));
        }
        esperarACola(Constantes.MAX_CONCURRENTES, Constantes.MAX_COLA);

        var rechazado = cliente.ejecutar(id(), new byte[8]);
        assertEquals(503, rechazado.codigo());
        assertEquals("5", rechazado.headers().get("retry-after"));
        assertEquals("RECHAZADA", MAPPER.readTree(rechazado.cuerpo()).get("resultado").asText());

        motor.soltar.countDown();
        for (Future<Integer> f : enVuelo) assertEquals(200, f.get(30, TimeUnit.SECONDS));
    }

    /** R3.3: saturacion no es enfermedad. Con la cola llena, /salud sigue en 200. */
    @Test
    @Timeout(60)
    void r33_conLaColaLlenaSaludSigueEn200() throws Exception {
        motor.bloquear = true;
        for (int i = 0; i < Constantes.MAX_CONCURRENTES + Constantes.MAX_COLA; i++) {
            hilos.submit(() -> cliente.ejecutar(id(), new byte[8]).codigo());
        }
        esperarACola(Constantes.MAX_CONCURRENTES, Constantes.MAX_COLA);

        var r = cliente.salud();
        assertEquals(200, r.codigo());
        JsonNode n = MAPPER.readTree(r.cuerpo());
        assertEquals("ok", n.get("daemon").asText());
        assertEquals(Constantes.MAX_CONCURRENTES, n.get("enVuelo").asInt());
        assertEquals(Constantes.MAX_COLA, n.get("enCola").asInt());
    }

    /** Seccion 3.4: si el daemon no responde el ping, /salud es 503. */
    @Test
    void saludEs503SiElDaemonNoResponde() throws Exception {
        motor.daemonVivo = false;
        var r = cliente.salud();
        assertEquals(503, r.codigo());
        assertEquals("caido", MAPPER.readTree(r.cuerpo()).get("daemon").asText());
    }

    @Test
    void saludEnReposo() throws Exception {
        JsonNode n = MAPPER.readTree(cliente.salud().cuerpo());
        assertEquals(0, n.get("enVuelo").asInt());
        assertEquals(0, n.get("enCola").asInt());
    }

    @Test
    void otraRutaEs404() throws Exception {
        assertEquals(404, cliente.enviar("GET /otra HTTP/1.1\r\nHost: localhost\r\n\r\n", null).codigo());
    }

    /** R11.7: al cerrar deja de aceptar y limpia los contenedores que quedaron. */
    @Test
    @Timeout(30)
    void apagadoOrdenado() throws Exception {
        servidor.close();
        assertTrue(motor.limpiado);
        assertThrows(IOException.class, () -> cliente.salud());
    }

    private void esperarACola(int enVuelo, int enCola) throws Exception {
        long limite = System.currentTimeMillis() + 30_000;
        while (System.currentTimeMillis() < limite) {
            JsonNode n = MAPPER.readTree(cliente.salud().cuerpo());
            if (n.get("enVuelo").asInt() == enVuelo && n.get("enCola").asInt() == enCola) return;
            Thread.sleep(20);
        }
        fail("la cola nunca llego a " + enVuelo + "/" + enCola);
    }
}
